meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2 ]
+collectNeededInputsFromObservation (ValueLE value1 value2) =
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2 ]
+collectNeededInputsFromObservation (ValueEQ value1 value2) =
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2 ]
+collectNeededInputsFromObservation (TrueObs) = S.empty
+collectNeededInputsFromObservation (FalseObs) = S.empty
+
+collectNeededInputsFromContract :: Contract -> S.Set IdInput
+collectNeededInputsFromContract Null = S.empty
+collectNeededInputsFromContract (Commit _ _ _ value _ _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromValue value
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Pay _ _ _ value _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromValue value
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Both contract1 contract2) =
+  S.unions [ collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Choice observation contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (When observation _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (While observation _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Scale value1 value2 value3 contract) = 
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2
+           , collectNeededInputsFromValue value3
+           , collectNeededInputsFromContract contract ]
+collectNeededInputsFromContract (Let _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Use _) = S.empty
+
+-- Add inputs and action ids to state.
+-- Return Nothing on redundant or irrelevant inputs
+addAnyInput :: BlockNumber -> AnyInput -> S.Set IdInput -> State -> Maybe State
+--  current block ---^
+addAnyInput _ (Action idInput) _ (State state)
+  | idInput `S.member` state.usedIds = Nothing
+  | true = Just (State (state { usedIds = S.insert idInput usedIdsSet }))
```

Comment:
Again, this might be neater with a lensy little `ix idInput`.

Reply:
:+1: